### PR TITLE
Switch to Node.js v18 + v20

### PR DESCRIPTION
**Pull request by Rugvip**
Just a draft for now, to check how this behaves in CI. It's still a while before we'll switch over

Current known issues:

- mock-fs breakage in v20: https://github.com/tschaub/mock-fs/issues/377

Update: Node.js v16 is now EOL so we're accelerating this with less optimal workarounds. Proper fix is pending in https://github.com/tschaub/mock-fs/pull/378

### CODE REVIEW

**Review by camilaibs — APPROVED** [review]

**Comment by freben** [comment]
got a conflict now

File: .changeset/silent-years-bake.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+'@backstage/plugin-scaffolder-backend-module-rails': patch
+'@techdocs/cli': patch
+'@backstage/cli': patch
+'@backstage/cli-common': patch
+'@backstage/create-app': patch
+'@backstage/codemods': patch
+'@backstage/repo-tools': patch
+---
+
+Bumped `@types/node` dependency to `^18.17.8`.
```
Comment (freben) [inline .changeset/silent-years-bake.md:11]: There's also the mock-fs bump here and there, is that less worth mentioning than the type bump? Not sure

**Review by freben — APPROVED** [review]

**Review by jhaals — APPROVED** [review]

**Comment by bforbis** [comment]
I ended up closing the linked issue #20001 as it seems this work is already in progress. Dropping the important details here as it seems relevant.

>I did some investigation and found that the current dependencies do not yet have packages available to upgrade to which will resolve this problem. This is blocked on https://github.com/Azure/azure-sdk-for-js/issues/26886 which will release a new major version of @azure/identity which can then pull in the latest version of @azure/msal-node which then updates the NodeJS engine restriction.
>
>The latest comments there says that this new package will likely be released in either October / November.

This will need to be pulled into the package `@backstage/integration`

**Comment by Rugvip** [comment]
Regarding the azure SDK issue, we're not seeing that here because we're on Yarn 3, which has removed the engines check.

This does mean that some users will hit this issue in moving to Node.js 20, but I don't think it's a reason for us not to move up to 20, especially with 16 no longer being supported.

**Pull request merged**